src/syntax/sugar.rkt

```diff
@@ -135,97 +140,218 @@
-         (loop body
-               (match (dict-ref props* ':precision #f)
-                 [#f ctx]
-                 [prec (struct-copy context ctx [repr (get-representation prec)])]))]
-        [`(cast ,body)
-         (define repr (context-repr ctx))
-         (define-values (body* repr*) (loop body ctx))
-         (if (equal? repr* repr) ; check if cast is redundant
-             (values body* repr)
-             (values (list (get-cast-impl repr* repr) body*) repr))]
-        [`(,(? constant-operator? x))
-         (define cnst (get-parametric-constant x (context-repr ctx)))
-         (values (list cnst) (impl-info cnst 'otype))]
-        [(list 'neg arg) ; non-standard but useful
-         (define-values (arg* atype) (loop arg ctx))
-         (define op* (get-parametric-operator 'neg atype))
-         (values (list op* arg*) (impl-info op* 'otype))]
-        [`(,op ,args ...)
-         (define-values (args* atypes) (for/lists (args* atypes) ([arg args]) (loop arg ctx)))
-         ;; Match guaranteed to succeed because we ran type-check first
-         (define op* (apply get-parametric-operator op atypes))
-         (values (cons op* args*) (impl-info op* 'otype))]
-        [(? variable?) (values expr (context-lookup ctx expr))]
-        [(? number?)
-         (define prec (representation-name (context-repr ctx)))
-         (define num
-           (match expr
-             [(or +inf.0 -inf.0 +nan.0) expr]
-             [(? exact?) expr]
-             [_ (inexact->exact expr)]))
-         (values (literal num prec) (context-repr ctx))])))
-  expr*)
+  (let loop ([expr (expand-expr prog)]
+             [prop-dict (repr->prop (context-repr ctx))])
+    (match expr
+      [(? number? n)
+       (literal (match n
+                  [(or +inf.0 -inf.0 +nan.0) expr]
+                  [(? exact?) expr]
+                  [_ (inexact->exact expr)])
+                (dict-ref prop-dict ':precision))]
+      [(? variable?) expr]
+      [(list 'if cond ift iff)
+       (define cond* (loop cond prop-dict))
+       (define ift* (loop ift prop-dict))
+       (define iff* (loop iff prop-dict))
+       (list 'if cond* ift* iff*)]
+      [(list '! props ... body) (loop body (apply dict-set prop-dict props))]
+      [(list 'neg arg) ; non-standard but useful [TODO: remove]
+       (define arg* (loop arg prop-dict))
+       (fpcore->impl-app '- prop-dict (list arg*) ctx)]
+      [(list 'cast arg) ; special case: unnecessary casts
+       (define arg* (loop arg prop-dict))
+       (define repr (get-representation (dict-ref prop-dict ':precision)))
+       (if (equal? (repr-of arg* ctx) repr) arg* (fpcore->impl-app 'cast prop-dict (list arg*) ctx))]
+      [(list op args ...)
+       (define args* (map (lambda (arg) (loop arg prop-dict)) args))
+       (fpcore->impl-app op prop-dict args* ctx)])))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; LImpl -> FPCore
+;; Translates from LImpl to an FPCore
+
```

Comment:
Ok, uhh, what's all this? Why do we need to turn it into an instruction vector / deduplicate common subexpressions?